### ISSUE TITLE
Enum `fromStringFuzzy` should trim() and replace inner whitespace with '_'

### ISSUE THREAD

**Issue by stefan-zh** [dropwizard/dropwizard#1985]
I think that the `Enums` class in `io.dropwizard.util` should `trim()` whitespace around the String and then convert inner whitespace to `_`. This way you can use space characters as delimiters too.
```java
final String text = value.trim()
            .replaceAll("\\s+", "_") // <- space can be used as a delimiter too
            .replace('-', '_')
            .replace('.', '_');
```

Somewhere before version `1.0` or `1.1` of Dropwizard it was possible to use the `fromString()` method on Enums to convert Strings to the right Enum, but not they all go through `Enums.fromStringFuzzy()`, which eliminates any custom conversion logic.

That would be a different issue, but wouldn't it be helpful to check whether the custom `Enum` class actually implements a `fromString()` and only in case it doesn't, fall back to `final Enum<?> constant = Enums.fromStringFuzzy(value, constants);` I guess this logic would go somewhere in `FuzzyEnumParamConverterProvider`

**Comment by stefan-zh** [comment]
Proposing this solution: https://github.com/dropwizard/dropwizard/pull/1987

**Comment by BorisNaguet** [comment]
Hello,

We have the same issue here, as we have an enum with a custom fromString() method. It's useful when the strings contains chars that are not allowed in java enum name for example.

It used to be managed by 
`org.glassfish.jersey.server.internal.inject.ParamConverters$TypeFromStringEnum`
that is included in 
`org.glassfish.jersey.server.internal.inject.ParamConverters$AggregatedProvider`

AggregatedProvider is still present but is shadowed by FuzzyEnumParamConverterProvider because `@Custom` providers have a higher priority.

**Comment by BorisNaguet** [comment]
I also noticed that all ConverterProviders are `@Singleton` except `FuzzyEnumParamConverterProvider`.

I don't think there's a reason for that?

**Comment by arteam** [comment]
I think it's a good idea. I can imagine a situation when we want to map enums to values with spaces in the middle (e.g. country names).

**Comment by arteam** [comment]
Closed via #2161
